csi/controller.go

```diff
@@ -230,12 +240,83 @@ func (s *OsdCsiServer) ValidateVolumeCapabilities(
 	return resp, nil
 }
 
+// ListVolumes is a CSI API which returns to the caller all volume ids
+// on this cluster. This includes ids created by CSI and ids created
+// using other interfaces. This is important because the user could
+// be requesting to mount a OSD volume created using non-CSI interfaces.
+//
+// This call does not yet implement tokens to due the following
+// issue: https://github.com/container-storage-interface/spec/issues/138
+func (s *OsdCsiServer) ListVolumes(
+	ctx context.Context,
+	req *csi.ListVolumesRequest,
+) (*csi.ListVolumesResponse, error) {
+
+	// Future: Once CSI is released, check version
+	// for now, just log it.
+	dlog.Debugf("ListVolumes req[%#v]", req)
+
+	// REQUIRE: Must at least that version is provided
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+
+	// Until the issue #138 on the CSI spec is resolved we will not support
+	// tokenization
+	if req.GetMaxEntries() != 0 {
+		return nil, status.Error(
+			codes.Unimplemented,
+			"Driver does not support tokenization. Please see "+
+				"https://github.com/container-storage-interface/spec/issues/138")
+	}
+
+	volumes, err := s.driver.Enumerate(&api.VolumeLocator{}, nil)
+	if err != nil {
+		errs := fmt.Sprintf("Unable to get list of volumes: %s", err.Error())
+		dlog.Errorln(errs)
+		return nil, status.Error(codes.Internal, errs)
+	}
+	entries := make([]*csi.ListVolumesResponse_Result_Entry, len(volumes))
+	for i, v := range volumes {
+		// Initialize entry
+		entries[i] = &csi.ListVolumesResponse_Result_Entry{
+			VolumeInfo: &csi.VolumeInfo{},
+		}
+
+		// Required
+		entries[i].VolumeInfo.Id = v.Id
+
+		// This entry is optional in the API, but OSD has
+		// the information available to provide it
+		entries[i].VolumeInfo.CapacityBytes = v.Spec.Size
+
+		// Attributes. We can add or remove as needed since they
+		// are optional and opaque to the CO but could be used
+		// for debuggig using a csi complient client.
+		entries[i].VolumeInfo.Attributes = map[string]string{
```

Comment:
We could add other attributes here as well:
- Parent 
- Secure

Reply:
Good point. I will add this.